Index: qemu/hw/i386/pc_piix.c
===================================================================
--- qemu.orig/hw/i386/pc_piix.c
+++ qemu/hw/i386/pc_piix.c
@@ -460,7 +460,6 @@ static QEMUMachine pc_i440fx_machine_v2_
     .name = "pc-i440fx-2.1",
     .alias = "pc",
     .init = pc_init_pci,
-    .is_default = 1,
 };
 
 #define PC_I440FX_2_0_MACHINE_OPTIONS PC_I440FX_2_1_MACHINE_OPTIONS
@@ -920,3 +919,37 @@ static void pc_machine_init(void)
 }
 
 machine_init(pc_machine_init);
+
+/* Ubuntu machine types */
+
+static QEMUMachine pc_machine_trusty = {
+	PC_DEFAULT_MACHINE_OPTIONS,
+	.default_machine_opts = "firmware=bios-256k.bin",
+	.name = "pc-i440fx-trusty",
+	.desc = "Ubuntu 14.04 PC (i440FX + PIIX, 1996)",
+	.hot_add_cpu = pc_hot_add_cpu,
+	.init = pc_init_pci_2_0,
+	.compat_props = (GlobalProperty[]) {
+		PC_COMPAT_2_0,
+		{ /* end of list */ }
+	},
+};
+
+static QEMUMachine pc_machine_utopic = {
+	PC_DEFAULT_MACHINE_OPTIONS,
+	.default_machine_opts = "firmware=bios-256k.bin",
+	.name = "pc-i440fx-utopic",
+	.desc = "Ubuntu 14.10 PC (i440FX + PIIX, 1996)",
+	.alias = "ubuntu",
+	.hot_add_cpu = pc_hot_add_cpu,
+	.init = pc_init_pci,
+	.is_default = 1,
+};
+
+static void ubuntu_machine_init(void)
+{
+	qemu_register_pc_machine(&pc_machine_trusty);
+	qemu_register_pc_machine(&pc_machine_utopic);
+}
+
+machine_init(ubuntu_machine_init);
