Index: qemu/configure
===================================================================
--- qemu.orig/configure	2013-05-22 19:56:19.781312969 +0200
+++ qemu/configure	2013-05-22 19:56:43.441193884 +0200
@@ -2662,6 +2662,14 @@
 #include <fcntl.h>
 #include <sys/stat.h>
 
+#if defined __stub_futimens || defined __stub___futimens
+choke me
+#endif
+
+#if defined __stub_utimensat || defined __stub___utimensat
+choke me
+#endif
+
 int main(void)
 {
     utimensat(AT_FDCWD, "foo", NULL, 0);
