From 25eaf329b13155dec0b65eb92cfb975781790206 Mon Sep 17 00:00:00 2001
From: Peter Maydell <peter.maydell@linaro.org>
Date: Tue, 4 Dec 2012 14:10:06 +0000
Subject: [PATCH 21/80] i2c-fixes: pull out omap-3-rev fixes

---
 hw/omap_i2c.c | 52 ++++++++++++++++++++++++++++++++++++++++------------
 1 file changed, 40 insertions(+), 12 deletions(-)

diff --git a/hw/omap_i2c.c b/hw/omap_i2c.c
index c432e54..6c991ed 100644
--- a/hw/omap_i2c.c
+++ b/hw/omap_i2c.c
@@ -337,10 +337,15 @@ static void omap_i2c_write(void *opaque, hwaddr addr,
         OMAP_RO_REG(addr);
         break;
     case 0x04:	/* I2C_IE */
-        if (s->revision >= OMAP3_INTR_REV)
+        if (s->revision < OMAP2_INTR_REV) {
+            s->mask = value & 0x1f;
+        } else if (s->revision < OMAP3_INTR_REV) {
+            s->mask = value & 0x3f;
+        } else if (s->revision == OMAP3_INTR_REV) {
             s->mask = value & 0x63ff;
-        else
-            s->mask = value & (s->revision < OMAP2_INTR_REV ? 0x1f : 0x3f);
+        } else { /* omap3630 */
+            s->mask = value & 0x6fff;
+        }
         omap_i2c_interrupts_update(s);
         break;
     case 0x08:	/* I2C_STAT */
@@ -348,16 +353,26 @@ static void omap_i2c_write(void *opaque, hwaddr addr,
             OMAP_RO_REG(addr);
         else {
             /* RRDY and XRDY are reset by hardware. (in all versions???) */
-            s->stat &= ~(value & (s->revision < OMAP3_INTR_REV ? 0x27 : 0x63e7));
+            if (s->revision < OMAP3_INTR_REV) {
+                value &= 0x27;
+            } else if (s->revision == OMAP3_INTR_REV) {
+                value &= 0x63e7;
+            } else { /* omap3630 */
+                value &= 0x6ee7;
+            }
+            s->stat &= ~value;
             omap_i2c_interrupts_update(s);
         }
         break;
 
     case 0x0c: /* I2C_IV / I2C_WE */
-        if (s->revision < OMAP3_INTR_REV)
+        if (s->revision < OMAP3_INTR_REV) {
             OMAP_RO_REG(addr);
-        else
+        } else if (s->revision == OMAP3_INTR_REV) {
             s->we = value & 0x636f;
+        } else { /* omap3630 */
+            s->we = value & 0x6f6f;
+        }
         break;
 
     case 0x14:	/* I2C_BUF */
@@ -481,20 +496,33 @@ static void omap_i2c_write(void *opaque, hwaddr addr,
         break;
 
     case 0x3c:	/* I2C_SYSTEST */
-        value &= s->revision < OMAP3_INTR_REV ? 0xf805 : 0xf815;
+        if (s->revision < OMAP3_INTR_REV) {
+            value &= 0xf805;
+        } else if (s->revision == OMAP3_INTR_REV) {
+            value &= 0xf815;
+        } else { /* omap3630 */
+            value = (value & 0xf835) | 0x1c00;
+        }
         if ((value & (1 << 15))) { /* ST_EN */
             fprintf(stderr, "%s: System Test not supported\n",
                     __FUNCTION__);
             s->test = (s->test & 0x0a) | value;
-        } else
-            s->test = (s->test & 0x1f) | (value & 0xf800);
-        if (value & (1 << 11))					/* SBB */
+        } else {
+            value &= ~0xff;
+            s->test = (s->test & 0x1f) | value;
+        }
+        if (value & (1 << 11)) { /* SBB */
             if (s->revision >= OMAP2_INTR_REV) {
                 s->stat |= 0x3f;
-                if (s->revision >= OMAP3_INTR_REV)
-                    s->stat |= 0x600;
+                if (s->revision >= OMAP3_INTR_REV) {
+                    s->stat |= 0x6300;
+                    if (s->revision > OMAP3_INTR_REV) {
+                        s->stat |= 0x0c00;
+                    }
+                }
                 omap_i2c_interrupts_update(s);
             }
+        }
         break;
 
     case 0x44: /* I2C_OA1 */
-- 
1.8.0

