From 8bebfa69b42f893e290f050d08dfadaf4fa30e91 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Juha=20Riihim=C3=A4ki?= <juha.riihimaki@nokia.com>
Date: Tue, 4 Dec 2012 14:10:08 +0000
Subject: [PATCH 32/80] fix omap_spi rxs flag status update
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Juha Riihimäki <juha.riihimaki@nokia.com>
---
 hw/omap_spi.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/hw/omap_spi.c b/hw/omap_spi.c
index 7bcf692..08d2462 100644
--- a/hw/omap_spi.c
+++ b/hw/omap_spi.c
@@ -229,8 +229,11 @@ static void omap_mcspi_transfer_run(OMAPSPIBusState *s, int chnum)
     ch->tx = 0;
     ch->status |= 1 << 2;               /* EOT */
     ch->status |= 1 << 1;               /* TXS */
-    if (trm != 2)
+    if (trm != 2) {
         ch->status |= 1;                /* RXS */
+    } else {
+        ch->status &= ~1;               /* RXS */
+    }
 
 intr_update:
     if ((ch->status & 1) &&	trm != 2 &&                     /* RXS */
-- 
1.8.0

