From fd945fe0e8bcbe8ee460a73166c8cc7468db5b72 Mon Sep 17 00:00:00 2001
From: Peter Maydell <peter.maydell@linaro.org>
Date: Tue, 4 Dec 2012 14:10:14 +0000
Subject: [PATCH 53/77] hw/nseries.c: Drop QEMU_VERSION from atags

(Why?)
---
 hw/nseries.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/hw/nseries.c b/hw/nseries.c
index 2549b82..6d0265d 100644
--- a/hw/nseries.c
+++ b/hw/nseries.c
@@ -1407,8 +1407,7 @@ static int n8x0_atag_setup(void *p, int model)
     stw_raw(w ++, 24);				/* u16 len */
     strcpy((void *) w, "hw-build");		/* char component[12] */
     w += 6;
-    strcpy((void *) w, "QEMU ");
-    pstrcat((void *) w, 12, qemu_get_version()); /* char version[12] */
+    strcpy((void *) w, "QEMU");		/* char version[12] */
     w += 6;
 
     tag = (model == 810) ? "1.1.10-qemu" : "1.1.6-qemu";
-- 
1.8.1.2

