From 9a7c0b981eec85a26a6cbca30410c708a6858b84 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Juha=20Riihim=C3=A4ki?= <juha.riihimaki@nokia.com>
Date: Mon, 18 Feb 2013 16:58:31 +0000
Subject: [PATCH 47/69] dsi: generate complexio ulps0 irq when asserting lane
 sig2
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Juha Riihimäki <juha.riihimaki@nokia.com>
---
 hw/display/omap_dss.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/hw/display/omap_dss.c b/hw/display/omap_dss.c
index d8fafa3..d328083 100644
--- a/hw/display/omap_dss.c
+++ b/hw/display/omap_dss.c
@@ -2329,6 +2329,11 @@ static void omap_dsi_write(void *opaque, hwaddr addr,
         TRACEDSI("DSI_COMPLEXIO_CFG_2 = 0x%08x", value);
         s->dsi.complexio_cfg2 = (value & 0xfffcffff)
                                 | (s->dsi.complexio_cfg2 & (3 << 16));
+        if (((value >> 5) & 3)) {
+            s->dsi.complexio_irqst |= (1 << 30); /* ULPSACTIVENOT_ALL0_IRQ */
+            s->dsi.irqst |= (1 << 10);           /* COMPLEXIO_ERR_IRQ */
+            omap_dss_interrupt_update(s);
+        }
         break;
     case 0x080: /* DSI_VM_TIMING4 */
         TRACEDSI("DSI_VM_TIMING4 = 0x%08x", value);
-- 
1.8.1.2

