From f0cb4feb25121fb8e1709892074cdd2543f03639 Mon Sep 17 00:00:00 2001
From: Peter Maydell <peter.maydell@linaro.org>
Date: Mon, 18 Feb 2013 16:58:32 +0000
Subject: [PATCH 52/69] hw/nseries.c: Force pointer visibility on n8x0

---
 hw/arm/nseries.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/hw/arm/nseries.c b/hw/arm/nseries.c
index 306cb30..d47bdb0 100644
--- a/hw/arm/nseries.c
+++ b/hw/arm/nseries.c
@@ -1471,6 +1471,7 @@ static void n8x0_init(QEMUMachineInitArgs *args,
         n810_tsc_setup(s);
         n810_kbd_setup(s);
     }
+    cursor_hide = 0; // who wants to use touchscreen without a pointer?
     n8x0_spi_setup(s);
     n8x0_dss_setup(s);
     n8x0_cbus_setup(s);
-- 
1.8.1.2

