From 80d37cd70440d6460b226fce8ee5b514d44da2f4 Mon Sep 17 00:00:00 2001
From: Peter Maydell <peter.maydell@linaro.org>
Date: Mon, 18 Feb 2013 16:58:32 +0000
Subject: [PATCH 53/69] hw/nseries.c: Drop QEMU_VERSION from atags

(Why?)
---
 hw/arm/nseries.c | 3 +--
 1 file changed, 1 insertion(+), 2 deletions(-)

diff --git a/hw/arm/nseries.c b/hw/arm/nseries.c
index d47bdb0..88fab81 100644
--- a/hw/arm/nseries.c
+++ b/hw/arm/nseries.c
@@ -1403,8 +1403,7 @@ static int n8x0_atag_setup(void *p, int model)
     stw_raw(w ++, 24);				/* u16 len */
     strcpy((void *) w, "hw-build");		/* char component[12] */
     w += 6;
-    strcpy((void *) w, "QEMU ");
-    pstrcat((void *) w, 12, qemu_get_version()); /* char version[12] */
+    strcpy((void *) w, "QEMU");		/* char version[12] */
     w += 6;
 
     tag = (model == 810) ? "1.1.10-qemu" : "1.1.6-qemu";
-- 
1.8.1.2

