From dece59f1897075af77d762209722fba3443bd273 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Juha=20Riihim=C3=A4ki?= <juha.riihimaki@nokia.com>
Date: Mon, 18 Feb 2013 16:58:35 +0000
Subject: [PATCH 65/77] musb: add dummy support for ulpi pass-through registers
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Juha Riihimäki <juha.riihimaki@nokia.com>
---
 hw/usb/hcd-musb.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/hw/usb/hcd-musb.c b/hw/usb/hcd-musb.c
index 64e9e83..20fd5b1 100644
--- a/hw/usb/hcd-musb.c
+++ b/hw/usb/hcd-musb.c
@@ -65,6 +65,7 @@
 #define MUSB_HDRC_ULPI_REGDATA	0x74
 #define MUSB_HDRC_ULPI_REGADDR	0x75
 #define MUSB_HDRC_ULPI_REGCTL	0x76
+#define MUSB_HDRC_ULPI_RAWDATA  0x77
 
 /* Extended config & PHY control */
 #define MUSB_HDRC_ENDCOUNT	0x78	/* 8 bit */
@@ -1293,6 +1294,16 @@ static uint32_t musb_readb(void *opaque, hwaddr addr)
         ep = ((addr - MUSB_HDRC_FIFO) >> 2) & 0xf;
         return musb_read_fifo(s->ep + ep);
 
+    case MUSB_HDRC_ULPI_REGADDR:
+    case MUSB_HDRC_ULPI_REGDATA:
+    case MUSB_HDRC_ULPI_RAWDATA:
+        /* TODO */
+        return 0x00;
+
+    case MUSB_HDRC_ULPI_REGCTL:
+        /* TODO */
+        return 0x02;
+
     default:
         TRACE("unknown register 0x%02x", (int) addr);
         return 0x00;
@@ -1380,6 +1391,13 @@ static void musb_writeb(void *opaque, hwaddr addr, uint32_t value)
         musb_write_fifo(s->ep + ep, value & 0xff);
         break;
 
+    case MUSB_HDRC_ULPI_REGADDR:
+    case MUSB_HDRC_ULPI_REGCTL:
+    case MUSB_HDRC_ULPI_REGDATA:
+    case MUSB_HDRC_ULPI_RAWDATA:
+        /* TODO */
+        break;
+
     default:
         TRACE("unknown register 0x%02x", (int) addr);
         break;
-- 
1.8.1.2

